gson/src/main/java/com/google/gson/internal/bind/TypeAdapters.java

```diff
@@ -776,9 +777,14 @@ public void write(JsonWriter out, Locale value) throws IOException {
 
     public EnumTypeAdapter(Class<T> classOfT) {
       try {
-        for (T constant : classOfT.getEnumConstants()) {
+        for (Field field : classOfT.getDeclaredFields()) {
+          if (!field.isEnumConstant()) {
+            continue;
+          }
+          field.setAccessible(true);
```

Comment:
Why make the field accessible? Enum element fields are public so making them accessible should not be necessary.

Reply:
Without this line, unittests failed with following error.
```
Failed tests:   testCollectionOfEnumsDeserialization(com.google.gson.functional.EnumTest): AssertionError (GSON 2.8.6-SNAPSHOT): java.lang.IllegalAccessException: class com.google.gson.internal.bind.TypeAdapters$EnumTypeAdapter cannot access a member of class com.google.gson.functional.EnumTest$MyEnum with modifiers "public static final"
  testTopLevelEnumDeserialization(com.google.gson.functional.EnumTest): AssertionError (GSON 2.8.6-SNAPSHOT): java.lang.IllegalAccessException: class com.google.gson.internal.bind.TypeAdapters$EnumTypeAdapter cannot access a member of class com.google.gson.functional.EnumTest$MyEnum with modifiers "public static final"
  testClassWithEnumFieldSerialization(com.google.gson.functional.EnumTest): java.lang.IllegalAccessException: class com.google.gson.internal.bind.TypeAdapters$EnumTypeAdapter cannot access a member of class com.google.gson.functional.EnumTest$MyEnum with modifiers "public static final"
  testClassWithEnumFieldDeserialization(com.google.gson.functional.EnumTest): AssertionError (GSON 2.8.6-SNAPSHOT): java.lang.IllegalAccessException: class com.google.gson.internal.bind.TypeAdapters$EnumTypeAdapter cannot access a member of class com.google.gson.functional.EnumTest$MyEnum with modifiers "public static final"
  testCollectionOfEnumsSerialization(com.google.gson.functional.EnumTest): AssertionError (GSON 2.8.6-SNAPSHOT): java.lang.IllegalAccessException: class com.google.gson.internal.bind.TypeAdapters$EnumTypeAdapter cannot access a member of class com.google.gson.functional.EnumTest$MyEnum with modifiers "public static final"
  testTopLevelEnumSerialization(com.google.gson.functional.EnumTest): java.lang.IllegalAccessException: class com.google.gson.internal.bind.TypeAdapters$EnumTypeAdapter cannot access a member of class com.google.gson.functional.EnumTest$MyEnum with modifiers "public static final"

```

Reply:
Enum constants are public, but the enum class itself isn't necessarily. If it isn't, then you'll get `IllegalAccessException`.

My concern here is that if there is a SecurityManager, the `field.setAccessible(true)` call might fail, whereas the code using `classOfT.getEnumConstants()` would have worked. So we'll have fixed one problem (getting enum constants in ProGuarded code) but introduced another (getting enum constants when there is a SecurityManager).

I think the code here should read something like this:

```java
for (final Field field : classOfT.getDeclaredFields()) {
  if (!field.isEnumConstant()) {
    continue;
  }
  AccessController.doPrivileged(
      new PrivilegedAction<Void>() {
        @Override Void run() {
          field.setAccessible(true);
          return null;
        }
      });
  @SuppressWarnings("unchecked")
  T constant = (T) field.get(null);
```

Reply:
Fixed at 59a8aedb37ae35d4d2a4306c92efd463fffaaf69